modules/godot/databags/databag_timer.cpp

```diff
@@ -0,0 +1,79 @@
+#include "databag_timer.h"
+
+void TimersDatabag::_bind_methods() {
+	add_method("is_valid_timer_handle", &TimersDatabag::script_is_valid_timer_handle);
+	add_method("new_timer", &TimersDatabag::script_new_timer);
+	add_method("new_precise_timer", &TimersDatabag::script_new_precise_timer);
+	add_method("restart_timer", &TimersDatabag::script_restart_timer);
+	add_method("restart_precise_timer", &TimersDatabag::script_restart_precise_timer);
+	add_method("is_done", &TimersDatabag::script_is_done);
+	add_method("destroy_timer", &TimersDatabag::script_destroy_timer);
+	add_method("get_remaining_seconds", &TimersDatabag::script_get_remaining_seconds);
+	add_method("get_remaining_microseconds", &TimersDatabag::script_get_remaining_microseconds);
+	add_method("get_remaining_or_passed_microseconds", &TimersDatabag::script_get_remaining_or_passed_microseconds);
+}
+
+void TimersDatabag::set_now(const uint64_t p_now) {
+	now = p_now;
+}
+
+bool TimersDatabag::is_valid_timer_handle(const godex::TimerHandle p_timer_handle) const {
+	return timers.size() < p_timer_handle.timer_index && timers[p_timer_handle.timer_index].generation == p_timer_handle.generation && timers[p_timer_handle.timer_index].end_time > 0;
+}
+
+godex::TimerHandle TimersDatabag::new_timer(const real_t p_end_in_seconds) {
+	ERR_FAIL_COND_V_MSG(p_end_in_seconds < 0.0, godex::TimerHandle(), "Can't make a new timer that ends at the past. To create an already timed out event, just enter 0.0 as time to end.");
+	return new_precise_timer(static_cast<uint64_t>(p_end_in_seconds * 1'000'000.0));
+}
+
+godex::TimerHandle TimersDatabag::new_precise_timer(const uint64_t p_end_in_microseconds) {
+	ERR_FAIL_COND_V_MSG(now + p_end_in_microseconds == 0, godex::TimerHandle(), "Can't make a new timer that ends at 0, it's invalid.");
+	if (destroyed_timers.size() > 0) {
+		godex::TimerHandle finalHandle = destroyed_timers[destroyed_timers.size() - 1];
+		internal_restart_precise_timer(finalHandle, now + p_end_in_microseconds);
+		destroyed_timers.remove_unordered(destroyed_timers.size() - 1);
+		return finalHandle;
+	} else {
+		timers.push_back(godex::Timer(now + p_end_in_microseconds, 0));
+		return godex::TimerHandle(timers.size() - 1, 0);
+	}
+}
+
+void TimersDatabag::restart_timer(const godex::TimerHandle p_timer_handle, const real_t p_end_in_seconds) {
+	ERR_FAIL_COND_MSG(p_end_in_seconds < 0.0, "Can't make a new timer that ends at the past. To create an already timed out event, just enter 0.0 as time to end.");
+	restart_precise_timer(p_timer_handle, static_cast<uint64_t>(p_end_in_seconds * 1'000'000.0));
+}
+
+void TimersDatabag::restart_precise_timer(const godex::TimerHandle p_timer_handle, const uint64_t p_end_in_microseconds) {
+	ERR_FAIL_COND_MSG(now + p_end_in_microseconds == 0, "Can't make a new timer that ends at 0, it's invalid.");
```

Comment:
Yeah. It actually checks for the improbable case of something setting a timer at 0 at the frame 0. It might happen if the game starts paused on the iteration when we introduce the global pause. Maybe you want to keep it, or maybe we could initialize the now with one microsecond at databag construct time. What do you think?

Reply:
If someone set 0 at frame 0 it causes an issue other than creating an expired timer?

Reply:
Nope :P . Meybe we should make a comment at the docs or header of the function and be done with it.

Reply:
Yee, I would just say that passing 0 will just create expired (done) timers, which is expected anyway also. So this check can be removed and we can create expired timers, someone will find a use for it :rofl: